### PR TITLE
Add implementation of emscripten_memcpy_big based on bulk memory.

### PR DESCRIPTION
**Pull request by sbc100**
These new functions live in `libbulkmemory` which only gets included if bulk memory is enabled (either via `-mbulk-memory` directly or indirectly via `-pthread).

benchmark results for benchmark.test_memcpy_1mb:

```
 v8:                       mean: 1.666
 v8-bulkmemory:            mean: 1.598
 v8-standalone-bulkmemory: mean: 1.576
 v8-standalone:            mean: 3.197
```

Here we can see the that when bulk memory is enabled its at least as fast if not faster than the JS version.

v8-standalone doesn't have emscripten_memcpy_big at all is is much slower, as expected.  By adding `-mbulk-memory` the standalone version becomes just as fast as the non-standalone.

### CODE REVIEW

File: emcc.py [hunk 1]
```diff
@@ -3571,6 +3574,8 @@ def consume_arg_file():
       settings.DISABLE_EXCEPTION_CATCHING = 1
       settings.DISABLE_EXCEPTION_THROWING = 1
       settings.WASM_EXCEPTIONS = 0
+    elif arg == '-mbulk-memory':
```
Comment (tlively) [inline emcc.py:3577]: Do we need to do anything to handle `-mno-bulk-memory`?

File: system/lib/libc/emscripten_memset.c [hunk 2]
```diff
@@ -1,8 +1,9 @@
-// XXX EMSCRIPTEN ASAN: build an uninstrumented version of memset
-#if defined(__EMSCRIPTEN__) && defined(__has_feature)
-#if __has_feature(address_sanitizer)
-#define memset __attribute__((no_sanitize("address"))) emscripten_builtin_memset
```
Comment (tlively) [inline system/lib/libc/emscripten_memset.c:4]: What happened to `emscripten_bultin_memset`?

File: system/lib/standalone/standalone.c [hunk 3]
```diff
@@ -152,7 +152,7 @@ int emscripten_resize_heap(size_t size) {
 }
 
 double emscripten_get_now(void) {
-  return (1000 * clock()) / (double)CLOCKS_PER_SEC;
+  return (1000ll * clock()) / (double)CLOCKS_PER_SEC;
```
Comment (tlively) [inline system/lib/standalone/standalone.c:155]: Is this related?

Comment (sbc100) [reply system/lib/standalone/standalone.c:155]: This was needed to make the benchmarks pass in standalone mode.  Without this we get wrap around in the timings and emscripten_get_now starts returning negative numbers after a while.

File: emcc.py (hunk 1 above)
Comment (sbc100) [reply emcc.py:3577]: Since the new (internal) BULK_MEMORY defaults to false I don't think so.

File: system/lib/libc/emscripten_memcpy.c [hunk 4]
```diff
@@ -29,7 +29,7 @@ static void *__memcpy(void *restrict dest, const void *restrict src, size_t n) {
   unsigned char *block_aligned_d_end;
   unsigned char *d_end;
 
-#ifndef EMSCRIPTEN_STANDALONE_WASM
+#if !defined(EMSCRIPTEN_STANDALONE_WASM) || defined(__wasm_bulk_memory__)
```
Comment (kripken) [inline system/lib/libc/emscripten_memcpy.c:32]: Perhaps it's worth replacing all of `memcpy` if bulk memory is enabled? It can all be a single bulk memory operation in that case, which is smaller. Unless we have data showing it's less fast on small sizes or something like that?

File: emcc.py (hunk 1 above)
Comment (tlively) [reply emcc.py:3577]: What about handling e.g. `-mbulk-memory -mno-bulk-memory`?

File: system/lib/libc/emscripten_memset.c (hunk 2 above)
Comment (tlively) [reply system/lib/libc/emscripten_memset.c:4]: Yeah, most of the `emscripten_builtin_*` functions are used by the sanitizers since we don't have real interposition.

File: system/lib/libc/emscripten_memcpy.c (hunk 4 above)
Comment (sbc100) [reply system/lib/libc/emscripten_memcpy.c:32]: Perhaps we can do that as a followup..  this change is nice and compact because it simply replaces the exists `_big` JS functions with a native function.   Collecting data on whether to shrink or remove the threshold seems like it can be separate thing we can look into after this lands.

File: emcc.py (hunk 1 above)
Comment (sbc100) [reply emcc.py:3577]: Fair enough :)

**Review by tlively — APPROVED** [review]

File: system/lib/libc/emscripten_memcpy.c (hunk 4 above)
Comment (kripken) [reply system/lib/libc/emscripten_memcpy.c:32]: Ok, sounds good. Since there is a speed benefit here, it seems worth a temporary size tradeoff for it. But I am very optimistic that replacing all of memcpy will end up better later.

**Review by kripken — APPROVED** [review]

File: system/lib/libc/emscripten_memcpy.c (hunk 4 above)
Comment (sbc100) [reply system/lib/libc/emscripten_memcpy.c:32]: Yes, once bulk memory is the default we can/should just always use it here, and depend on binaryen to lower it away.  Until it that time though we don't want to depend on binaryen in the default case.

Comment (kripken) [reply system/lib/libc/emscripten_memcpy.c:32]: Ah, right, makes sense to wait for bulk memory to be on by default.

**Comment by haberman** [comment]
If bulk memory is available, why can't `memcpy()` and `memset()` use `memory.copy` and `memory.fill` exclusively? Why does `emscripten_memcpy_big()` need to exist at all in that case?

This code seems to be written with an assumption that the bulk memory operations have some kind of overhead that makes them only worthwhile for copies/fills of greater than 512 bytes.  Is this true?

**Comment by sbc100** [comment]
> If bulk memory is available, why can't `memcpy()` and `memset()` use `memory.copy` and `memory.fill` exclusively? Why does `emscripten_memcpy_big()` need to exist at all in that case?
> 
> This code seems to be written with an assumption that the bulk memory operations have some kind of overhead that makes them only worthwhile for copies/fills of greater than 512 bytes. Is this true?

I think you could well be correct and we might just want to completely remove the traditional `__musl_memcpy` when bulk memory is available.  

The code that this was replacing involved calling out the JavaScript which certain had/has a non-zero overhead.   We would probably want measure imperially weather `memory.copy` and `memory.fill` are faster or not for small copies, but it seems likely to me that there would have the same or better performance costs.

Do you have a particular reason to want this change?  For example, is the cost of including `__musl_memcpy` concerning you?

**Comment by haberman** [comment]
> Do you have a particular reason to want this change? For example, is the cost of including __musl_memcpy concerning you?

I am looking at a profile where `__memcpy()` (and `__musl_memset()` to a lesser extent) are significant costs, and I am hoping that using `memory.copy` and `memory.fill` would speed up my benchmark.

I don't know of a way to test this theory. Do you know of any way to force `memcpy()` to compile to `memory.copy`?  I tried using `__builtin_memcpy()`, but oddly that still seems to call the `__memcpy()`.

**Comment by sbc100** [comment]
The simplest way to do that would probably be to just remove the `if (n >= 512) {` check from memcpy.c, and then rebuild libc (either via `./embuilder build libc --force` or `./emcc --clear-cache` (to force all libraries to be rebuilt).
